apps/dashboard/src/main/java/com/akto/action/testing/MarketplaceAction.java

```diff
@@ -38,18 +41,65 @@ public String fetchTestingSources() {
     TestCategory category;
     Severity severity;
     String description;
+    List<String> tags;
+    String searchText;
+    List<TestSourceConfig> searchResults = new ArrayList<>();
+    private List<TestSubCategory> searchAktoTests;
+    private TestCategory[] categories;
+
     public String addCustomTest() {
         TestSourceConfig alreadyExists = TestSourceConfigsDao.instance.findOne("_id", url);
         if (alreadyExists != null) {
             addActionError("This test file has already been added");
             return ERROR.toUpperCase();            
         }
 
-        TestSourceConfig elem = new TestSourceConfig(url, category, subcategory, severity, description, getSUser().getLogin(), Context.now());
+        TestSourceConfig elem = new TestSourceConfig(url, category, subcategory, severity, description, getSUser().getLogin(), Context.now(),tags);
         TestSourceConfigsDao.instance.insertOne(elem);
         return Action.SUCCESS.toUpperCase();
     }
 
+    private void searchUtilityFunction(){
+        this.searchAktoTests = new ArrayList<>();
+        this.categories = GlobalEnums.TestCategory.values();
+        if(this.searchText == null){
+            for(TestSubCategory tsc : GlobalEnums.TestSubCategory.getValuesArray()){
+                this.searchAktoTests.add(tsc);
+            }
+            this.searchResults = TestSourceConfigsDao.instance.findAll(new BasicDBObject());
+        }
+
+        //fill from Updated test-source-config collection in mongodb
+        else{
+            this.searchResults = TestSourceConfigsDao.instance.findAll(Filters.or(
+                Filters.regex("severity", this.searchText, "i"),
+                Filters.regex("category", this.searchText, "i"),
+                Filters.regex("tags", this.searchText, "i"),
+                Filters.regex("description", this.searchText, "i"),
+                Filters.regex("subcategory", this.searchText, "i")
+            ));
+
+            this.searchText = this.searchText.toLowerCase();
+            //fill from akto tests in global enums
+            for(TestSubCategory tsc : GlobalEnums.TestSubCategory.getValuesArray()){
+                String testCategory = tsc.getSuperCategory().getName().toLowerCase();
```

Comment:
should we match on subcategory as well for inbuilt tests?

Reply:
there are no subcategories in inbuiltTests, still changed getName to getTestName for better search